### PR TITLE
[ssh-gateway] fix missing output when running simple command

### PR DESCRIPTION
**Pull request by iQQBot**
## Description

This PR fixes missing output in SSH gateway when running a simple command, and also upgrade golang crypto library

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at e9c1f91</samp>

Fixed security and compatibility issues in `ws-proxy` component. Refactored SSH proxy forwarding code to prevent crashes and handle errors better.

</details>

## Related Issue(s)

Fixes EXP-246

## How to test

1. start a workspace in preview environment
2. get ssh connection string via access token
3. run `watch -n 0.1 ssh -T 'workspaceID#token@workspace.url' 'date +%s%N'` test for 1minute
4. also run this command in gitpod.io workspace, and compare output.

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - pd-ssh-output</li>
	<li><b>🔗 URL</b> - <a href="https://pd-ssh-output.preview.gitpod-dev.com/workspaces" target="_blank">pd-ssh-output.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - pd-ssh-output-gha.14207</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] with-integration-tests=ssh
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
- [x] with-monitoring
</details>

/hold

### CODE REVIEW

**Review by aledbf — APPROVED** [review]

**Comment by akosyakov** [comment]
@iQQBot there is the actual fix? in the library?

**Comment by akosyakov** [comment]
ok, I tried with VS Code Desktop and Local SSH over direct SSH, it does not help there, it knows fails differently, it seems it can tunnel, but then interrupts, before could not forward at all.

```
[09:55:56.820] Remote server is listening on 36127
[09:55:56.820] Parsed server configuration: {"serverConfiguration":{"remoteListeningOn":{"port":36127},"osReleaseId":"ubuntu","arch":"x86_64","webUiAccessToken":"","sshAuthSock":"","display":"","tmpDir":"/tmp","platform":"linux","connectionToken":"1a1aa111-a1aa-111a-1111-aa1a1a1a111a"},"extInstallTime":239,"installUnpackCode":""}
[09:55:56.821] Persisting server connection details to /Users/[user]/Library/Application Support/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-0fdc06be-660393deaaa6d1996740ff4880f1bad43768c814-0.102.0/data.json
[09:55:56.826] Starting forwarding server. localPort 64392 -> socksPort 64384 -> remotePort 36127
[09:55:56.827] Forwarding server listening on 64392
[09:55:56.827] Waiting for ssh tunnel to be ready
[09:55:56.828] [Forwarding server 64392] Got connection 0
[09:55:56.829] Tunneled 36127 to local port 64392
[09:55:56.829] Resolved "ssh-remote+7b22686f73744e616d65223a22616b6f7379616b6f762d70617263656c64656d6f2d373476766964317a3762702e7673732e70642d7373682d6f75747075742e707265766965772e676974706f642d6465762e636f6d222c2275736572223a22616b6f7379616b6f762d70617263656c64656d6f2d373476766964317a376270227d" to "127.0.0.1:64392"
[09:55:56.837] ------




[09:55:56.866] [Forwarding server 64392] Got connection 1
[09:55:56.961] > local-server-2> ssh child died, shutting down
[09:55:56.961] Failed to set up socket for dynamic port forward to remote port 36127: Socket closed. Is the remote port correct?
[09:55:56.962] Failed to set up socket for dynamic port forward to remote port 36127: Socket closed. Is the remote port correct?
[09:55:56.966] Local server exit: 0
```

**Comment by akosyakov** [comment]
> not related this PR

What interesting though that before it won't ever reach `[Forwarding server 64392] Got connection 1` so this PR did fix initiation of the forwarding somehow.

**Comment by akosyakov** [comment]
@iQQBot Good to retest?

File: components/ws-proxy/pkg/sshproxy/forward.go [hunk 1]
```diff
@@ -53,32 +53,46 @@ func (s *Server) ChannelForward(ctx context.Context, session *Session, targetCon
...
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+
 	go func() {
-		io.Copy(targetChan, originChan)
-		targetChan.CloseWrite()
+		defer wg.Done()
+		_, _ = io.Copy(targetChan, originChan)
+		targetChannelWg.Done()
+		targetChannelWg.Wait()
```
Comment (akosyakov) [inline components/ws-proxy/pkg/sshproxy/forward.go:68]: I wonder should not we wait for stderr/stdout in both directions? won't close on one end trigger close on another

Comment (akosyakov) [reply components/ws-proxy/pkg/sshproxy/forward.go:68]: yeah, I meant there we can simplify with one wg, which waits for both directions, but like that is fine as well 👍

**Comment by akosyakov** [comment]
@iQQBot preview envs are down?

**Comment by mustard-mh** [comment]
> ok, I tried with VS Code Desktop and Local SSH over direct SSH, it does not help there, it now fails differently though, it seems it can tunnel, but then interrupts, before could not forward at all.
> 
> ```
> [09:55:56.820] Remote server is listening on 36127
> [09:55:56.820] Parsed server configuration: {"serverConfiguration":{"remoteListeningOn":{"port":36127},"osReleaseId":"ubuntu","arch":"x86_64","webUiAccessToken":"","sshAuthSock":"","display":"","tmpDir":"/tmp","platform":"linux","connectionToken":"1a1aa111-a1aa-111a-1111-aa1a1a1a111a"},"extInstallTime":239,"installUnpackCode":""}
> [09:55:56.821] Persisting server connection details to /Users/[user]/Library/Application Support/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-0fdc06be-660393deaaa6d1996740ff4880f1bad43768c814-0.102.0/data.json
> [09:55:56.826] Starting forwarding server. localPort 64392 -> socksPort 64384 -> remotePort 36127
> [09:55:56.827] Forwarding server listening on 64392
> [09:55:56.827] Waiting for ssh tunnel to be ready
> [09:55:56.828] [Forwarding server 64392] Got connection 0
> [09:55:56.829] Tunneled 36127 to local port 64392
> [09:55:56.829] Resolved "ssh-remote+7b22686f73744e616d65223a22616b6f7379616b6f762d70617263656c64656d6f2d373476766964317a3762702e7673732e70642d7373682d6f75747075742e707265766965772e676974706f642d6465762e636f6d222c2275736572223a22616b6f7379616b6f762d70617263656c64656d6f2d373476766964317a376270227d" to "127.0.0.1:64392"
> [09:55:56.837] ------
> 
> 
> 
> 
> [09:55:56.866] [Forwarding server 64392] Got connection 1
> [09:55:56.961] > local-server-2> ssh child died, shutting down
> [09:55:56.961] Failed to set up socket for dynamic port forward to remote port 36127: Socket closed. Is the remote port correct?
> [09:55:56.962] Failed to set up socket for dynamic port forward to remote port 36127: Socket closed. Is the remote port correct?
> [09:55:56.966] Local server exit: 0
> ```

@akosyakov  Found bug in https://github.com/microsoft/dev-tunnels-ssh that it omit `SSH_MSG_CHANNEL_EXTENDED_DATA` and throw error, will create a fix PR later today.

Which should make local-ssh x gateway works

**Comment by akosyakov** [comment]
> @akosyakov Found bug in https://github.com/microsoft/dev-tunnels-ssh that it omit SSH_MSG_CHANNEL_EXTENDED_DATA and throw error, will create a fix PR later today.

@mustard-mh I checked it is the same with the tunnel but it works. I wonder whether maybe it works with all changes together now. Although if we resolved localhost hanging then exploring direct SSH is not a priority.

**Comment by mustard-mh** [comment]
> @mustard-mh I checked it is the same with the tunnel but it works. I wonder whether maybe it works with all changes together now. Although if we resolved localhost hanging then exploring direct SSH is not a priority.

@akosyakov  It's because golang lib doesn't ignore `SSH_MSG_UNIMPLEMENTED ` as [rfc said](https://datatracker.ietf.org/doc/html/rfc4253#section-11.4:~:text=Such%20messages%20MUST%20be%20otherwise%20ignored) https://github.com/gitpod-io/golang-crypto/blob/master/ssh/messages.go#L844

When we use local-ssh x gateway, dev-tunnel will send `SSH_MSG_UNIMPLEMENTED` it doesn't resolve `SSH_MSG_CHANNEL_EXTENDED_DATA`.

So we can fix it in golang-crypto or dev-tunnel

**Pull request merged**